### ISSUE TITLE
Question on running iscsi-gateway containers

### ISSUE THREAD

**Issue by Venator-Fox** [ceph/ceph-container#1214]
Hello, sorry if this is the wrong place to post.. super basic questions..
I've spent the better part of the day attempting to get a 2nd gateway added.
Whenever I perform the following from any host:

~~~
<...>
o- gateways ....................................... [Up: 1/1, Portals: 1]
  o- someFirstHost ................................ [172.31.10.100 (UP)]

/iscsi-target...-igw/gateways> create someSecondHost 172.31.10.101
Adding gateway, sync'ing 0 disk(s) and 0 client(s)
Failed : 500 INTERNAL SERVER ERROR
~~~

... I get `500 INTERNAL SERVER ERROR` returned

Attempting to go to `someSecondHost` and running the same motions in `gwcli` returns the `500 INTERNAL SERVER ERROR` as well.

Here are the first few lines from `gwcli -d`

~~~
CMD: ../gateways/ create someSecondHost 172.31.10.101 nosync=False skipchecks=false
Adding gateway, sync'ing 0 disk(s) and 0 client(s)
Failed API request: PUT http://127.0.0.1:5000/api/gateway/someSecondHost
<...>
    <title>ValueError: No JSON object could be decoded // Werkzeug Debugger</title>
<...>
~~~

Did a `curl` on `http://127.0.0.1:5000/api` and it returns. stuff and things. Doing it on, from, and to the remote hosts returns data as well.

The containers were started based on the ceph-ansible templates, but I added `--net=host`

~~~
docker run -d -v /etc/localtime:/etc/localtime:ro --privileged --cap-add=ALL --net=host -v /dev:/dev -v /lib/modules:/lib/modules -v /etc/ceph:/etc/ceph --name=rbd-target-api ceph/daemon rbd_target_api
docker run -d -v /etc/localtime:/etc/localtime:ro --privileged --cap-add=ALL --net=host -v /dev:/dev -v /lib/modules:/lib/modules -v /etc/ceph:/etc/ceph --name=rbd-target-gw ceph/daemon rbd_target_gw
docker run -d -v /etc/localtime:/etc/localtime:ro --privileged --cap-add=ALL --net=host -v /dev:/dev -v /lib/modules:/lib/modules -v /etc/ceph:/etc/ceph --name=tcmu-runner ceph/daemon tcmu_runner
~~~

I'm running `gwcli` in the rbd_target_api container and I'm 95.5% sure I'm doing everything wrong.

The contents of `iscsi-gateway.cfg` in all hosts are basically an exact paste of the block provided in `http://docs.ceph.com/docs/mimic/rbd/iscsi-target-cli/`, with `trusted_ip_list` uncommented with my host IP's

Any tips would be appreciated. Thanks!

**Environment**:
* OS (e.g. from /etc/os-release): CentOS Linux 7
* Kernel (e.g. `uname -a`): 3.10.0-862.14.4.el7.x86_64
* Docker version (e.g. `docker version`): 1.13.1
* Ceph version (e.g. `ceph -v`): 13.2.2

**Comment by leseb** [comment]
@mikechristie can you look at the iscsi portion please?

**Comment by mikechristie** [comment]
I do not think the the container support works. The tools/components get installed to containers, but they are not usable as is.

1. The rbd-target-api, rbd-target-gw and tcmu-runner containers need configfs mounted at /sys/kernel/config.

Was ansible docker supposed to do that or did you want me to modify apps running in those containers to do that themselves? For gluster iscsi the equivalent of ansible/docker do it so we had not modified the tools.

2. The tasks in configure_iscsi.yml are not run for the container case, so the objects in iscsigws.yml, like the targets, luns and clients, are not setup.

Those tasks need ceph-iscsi-config installed so do we need to run it from the rbd-target-api or rbd-target-gw container.

3. infrastructure-playbooks/purge-iscsi-gateways.yml has a similar problem as #2 where it needs ceph-iscsi-config.

**Comment by mikechristie** [comment]
@Venator-Fox 

To verify you are hitting issue #1, you can do:

mount -t configfs none /sys/kernel/config

in the rbd-target-api, rbd-target-gw and tcmu-runner containers. The gwcli commands should then work.

**Comment by Venator-Fox** [comment]
@mikechristie Yes, that seems to have resolved the errors with `gwcli`. Thank you.

Am I using these containers improperly? There is an oddity that initiators are unable to see any devices. Not sure if this is caused my improper use of these containers.

Is that what you were referring to with `#2`?

~~~
  o- iscsi-target .................................. [Targets: 1]
    o- iqn.2018-10.redhat.iscsi-gw:iscsi-igw ....... [Gateways: 4]
      o- gateways .................................. [Up: 4/4, Portals: 4]
      | o- someFirstHost............................ [172.31.10.100 (UP)]
      | o- someSecondHost .......................... [172.31.10.101 (UP)]
      | o- someThirdHost ........................... [172.31.10.102 (UP)]
      | o- someFourthHost .......................... [172.31.10.103 (UP)]
      o- host-groups ............................... [Groups : 0]
      o- hosts ..................................... [Hosts: 2]
        o- iqn.2018-10.com.redhat:rhelclienttest ... [LOGGED-IN, Auth: CHAP, Disks: 1(32G)]
        | o- lun 0 ................................. [rbd.disk_1(32G), Owner: someFirstHost]
        o- iqn.2018-10.com.microsoft:2k16clienttest. [LOGGED-IN, Auth: CHAP, Disks: 1(128G)]
          o- lun 0 ................................. [rbd.disk_2(128G), Owner: someFourthHost]
~~~

**Comment by mikechristie** [comment]
For my #2, I was just saying that for the non-container ansible setup you can add targets, luns, etc to the iscsigws.yml and when you run the ansible playbook command they will be created for you. For the container case you cannot.

It looks like from your gwcli output you worked around #2 by creating them with gwcli which is valid. With the target setup in your above output it looks like everything should be ok and the initiators are logged in so it should be working ok.

For the RHEL client if you run

iscsiadm -m session -P 3

on the initiator system you can see the session is logged in right? If you do

iscsiadm -m session --rescan

do you see any errors in /var/log/messages on the initiator system?

**Comment by Venator-Fox** [comment]
Performed a fresh install of RHEL7.5 and followed docs for multipath, logged in.

~~~
[root@rhelclienttest ~]# iscsiadm -m node -T iqn.2018-10.redhat.iscsi-gw:iscsi-igw -l
Logging in to [iface: default, target: iqn.2018-10.redhat.iscsi-gw:iscsi-igw, portal: 172.31.10.100,3260] (multiple)
Logging in to [iface: default, target: iqn.2018-10.redhat.iscsi-gw:iscsi-igw, portal: 172.31.10.101,3260] (multiple)
Logging in to [iface: default, target: iqn.2018-10.redhat.iscsi-gw:iscsi-igw, portal: 172.31.10.102,3260] (multiple)
Logging in to [iface: default, target: iqn.2018-10.redhat.iscsi-gw:iscsi-igw, portal: 172.31.10.103,3260] (multiple)
Login to [iface: default, target: iqn.2018-10.redhat.iscsi-gw:iscsi-igw, portal: 172.31.10.100,3260] successful.
Login to [iface: default, target: iqn.2018-10.redhat.iscsi-gw:iscsi-igw, portal: 172.31.10.101,3260] successful.
Login to [iface: default, target: iqn.2018-10.redhat.iscsi-gw:iscsi-igw, portal: 172.31.10.102,3260] successful.
Login to [iface: default, target: iqn.2018-10.redhat.iscsi-gw:iscsi-igw, portal: 172.31.10.103,3260] successful.
~~~

Nothing is returned after performing `multipath -ll`
No devices in `/dev/mapper` either

Ran `iscsiadm -m session -P 3` on this client system and verified it is logged in. `gwcli` on the target system also states logged in.

Output of `iscsiadm -m session --rescan`

~~~
Rescanning session [sid: 1, target: iqn.2018-10.redhat.iscsi-gw:iscsi-igw, portal: 172.31.210.100,3260]
Rescanning session [sid: 2, target: iqn.2018-10.redhat.iscsi-gw:iscsi-igw, portal: 172.31.210.101,3260]
Rescanning session [sid: 3, target: iqn.2018-10.redhat.iscsi-gw:iscsi-igw, portal: 172.31.210.102,3260]
Rescanning session [sid: 4, target: iqn.2018-10.redhat.iscsi-gw:iscsi-igw, portal: 172.31.210.103,3260]
~~~

I see no messages standing out in `/var/log/messages` other than it is operational.

Added a bunch of luns and rescanned again

~~~
o- iqn.2018-10.com.redhat:client .. [LOGGED-IN, Auth: CHAP, Disks: 8(480G)]
  o- lun 0 .. [rbd.disk_1(32G), Owner: someFirstHost]
  o- lun 1 .. [rbd.someDisk1(64G), Owner: someSecondHost]
  o- lun 2 .. [rbd.someDisk2(64G), Owner: someThirdHost]
  o- lun 3 .. [rbd.someDisk3(64G), Owner: someFourthHost]
  o- lun 4 .. [rbd.someDisk4(64G), Owner: someFirstHost]
  o- lun 5 .. [rbd.someDisk5(64G), Owner: someSecondHost]
  o- lun 6 .. [rbd.someDisk6(64G), Owner: someThirdHost]
  o- lun 7 .. [rbd.someDisk7(64G), Owner: someFourthHost]
~~~

Unable to find anything mapped

`multipath -v4` has a few errors, repeated 6 times, but I'm not sure its related.

~~~
<...>
Oct 24 11:33:04 rhelclienttest multipathd: dm-2: remove map (uevent)
Oct 24 11:38:25 rhelclienttest kernel: device-mapper: table: 253:2: multipath: error getting device
Oct 24 11:38:25 rhelclienttest kernel: device-mapper: ioctl: error adding target to table
Oct 24 11:38:25 rhelclienttest multipathd: dm-2: remove map (uevent)
<...>
~~~

Something is locked? Maybe because I have not blacklisted any devices.

I'll do more troubleshooting on MPIO. The multipath conf was a paste from the docs.

Same story with 2k16 mpio, logs in, but no devices.
Since I can't get any initiator to display devices (CentOS, RHEL, 2k16), I may have setup something wrong on the target system. I'll start over with the entire ceph cluster and try again, paying more attention to details to see if I catch anything.

**Comment by mikechristie** [comment]
The device-mapper errors are not related. If in /var/log/messages on the initiator side you are not seeing any scsi devices being found or iscsiadm -m session -P 3 does not show any disks/LUNs found then we are not even getting to the multipath setup stage yet.

I think something in the target side kernel is not right and it is misreporting the LUNs. Let me do some more debugging.

I am going on vacation until next Monday, so I will not be able t reply until then.

**Comment by mikechristie** [comment]
@Venator-Fox 

My target side theory was wrong. I tested the same version as you are running and it works ok for me.

Could you tell me the tcmu-runner version you have and in the tcmu-runner container could you do the following on one just one of the gw nodes:

1. Enable debugging. In /etc/tcmu/tcmu.conf set log_level to 5.
2. On the target node do:

echo "module iscsi_target_mod +p" > /sys/kernel/debug/dynamic_debug/control

3. On the RHEL initiator do

iscsiadm -m node --target $your-target --logout
iscsiadm -m node --target $your-target --login

and then send me the output from /var/log/tcmu-runner.log in the tcmu-runner container and the kernel output from /var/log/messages from the target node.

**Comment by mikechristie** [comment]
It looks like you just forgot to the the mount configfs command in the tcmu-runner container as well as the rbd-target-api one. Also make sure it is mounted in the rbd-target-gw container.

leeseb also just merged this patch:

https://github.com/ceph/ceph-container/commit/9f5e6730580d55ffedfcbe0fd1f16660ab6ea8f4

which will fix your issue.

**Comment by Venator-Fox** [comment]
Thank you for the patch.

`mount -t configfs none /sys/kernel/config` was run on all 3 containers. The message returned was that it was already mounted. Still unable to see any devices on initiator.. 

Since it works fine when you tested it, it's something on my end.
I'll rebuild the hosts, read more docs, and try again.

Thanks again for your help in resolving the original issue!

**Comment by mikechristie** [comment]
@Venator-Fox 

In the tcmu-runner container do you see:

/sys/kernel/config/target/core/user_$N/$name/attrib/hw_block_size

where $N is just a integer value and $name is the pool_name.image_name you set when your an th gwcli disk creation command?

**Comment by Venator-Fox** [comment]
Hi @mikechristie Sorry for the late reply, was out of home country.

Yes:
~~~
$ docker exec -ti ceph-rbd_target_api /bin/bash 
$ cat /sys/kernel/config/target/core/user_0/rbd.disk_0/attrib/hw_block_size
512
~~~

I will rebuild the container hosts tomorrow to see if that resolves.

**Comment by Venator-Fox** [comment]
Recreated everything on a fresh hosts (CentOS) and still having the same issue :/
Even disabled firewalld and SELinux this time around as well... for troubleshooting purpose.

***I know it explicitly says not to do this***, but for erm... troubleshooting purpose.. I used `targetcli` to create another lun in the same tpg (just to see if it shows). ***This lun indeed shows up on initiators.***

~~~
/iscsi/iqn.20...at:rh7-client> create tpg_lun_or_backstore=/backstores/block/sda mapped_lun=1
Created LUN 1.
Created Mapped LUN 1.
/iscsi/iqn.20...at:rh7-client> ls
o- iqn.1994-05.com.redhat:rh7-client ...... [1-way auth, Mapped LUNs: 2]
  o- mapped_lun0 .......................... [lun0 user/rbd.disk_1 (rw)]
  o- mapped_lun1 .......................... [lun1 block/sda (rw)]
/iscsi/iqn.20...at:rh7-client>
~~~

On the initiator, mapped_lun0 still does not show up, but mapped_lun1 does. Cannot figure out why rbd luns are invisible to initators. 


The following new containers are interpreted from <https://github.com/ceph/ceph-ansible/tree/master/roles/ceph-iscsi-gw/templates>:

~~~
docker run -d --restart always --name ceph-rbd_target_api -d --net=host --privileged --cap-add=ALL -v /etc/localtime:/etc/localtime:ro -v /lib/modules:/lib/modules -v /etc/ceph:/etc/ceph -v /dev:/dev -e CEPH_DAEMON=RBD_TARGET_API ceph/daemon:latest
docker run -d --restart always --name ceph-rbd_target_gw -d --net=host --privileged --cap-add=ALL -v /etc/localtime:/etc/localtime:ro -v /lib/modules:/lib/modules -v /etc/ceph:/etc/ceph -v /dev:/dev -e CEPH_DAEMON=RBD_TARGET_GW ceph/daemon:latest
docker run -d --restart always --name ceph-tcmu_runner -d --net=host --privileged --cap-add=ALL -v /etc/localtime:/etc/localtime:ro -v /lib/modules:/lib/modules -v /etc/ceph:/etc/ceph -v /dev:/dev -e CEPH_DAEMON=TCMU_RUNNER ceph/daemon:latest
~~~

**Comment by mikechristie** [comment]
Sorry for the late response. This got put in my spam filter.

It probably works because tcmu-runner is still not able to access some file in configfs.  In the last attempt we saw the error:

2018-10-24 18:56:33.515 21860 [ERROR] tcmu_get_cfgfs_int:33: Could not open configfs to read attribute /sys/kernel/config/target/core/user_0/rbd.disk_1/attrib/hw_block_size: No such file or directory

and that prevented all SCSI commands from being executed and so the initiator could not scan and detect the LUN.

target_core_iblock accessing sda does not need to access any configfs files to execute SCSI commands.

Can you send the /var/log/tcmu-runner.log again or check that in the tcmu-runner container that hw_block_size file exists?

**Comment by Venator-Fox** [comment]
It appears to be present:

~~~
$ docker exec -ti ceph-tcmu_runner /bin/bash
$ cat /sys/kernel/config/target/core/user_1/rbd.disk_2/attrib/hw_block_size 
512
~~~

Here are some logs from `/var/log/tcmu-runner.log`  

~~~
2018-11-20 10:12:50.997 97 [INFO] tcmu_set_log_level:98: log level now is DEBUG SCSI CMD
[DEBUG] create_file_output:434: Attempting to use '/var/log//tcmu-runner.log' as the log file path
<...>
2018-11-20 10:13:04.726 98 [DEBUG] load_our_module:529: Module 'target_core_user' is already loaded
<...>
2018-11-20 10:28:38.271 98 [DEBUG] dev_added:780 rbd/rbd.disk_2: Got block_size 512, size in bytes 34359738368
<...>
2018-11-20 10:28:38.282 98 [DEBUG] timer_check_and_set_def:391 rbd/rbd.disk_2: The cluster's default osd op timeout(0.000000), osd heartbeat grace(20) interval(6)
2018-11-20 10:28:38.295 98 [DEBUG] tcmu_rbd_detect_device_class:334 rbd/rbd.disk_2: Pool rbd using crush rule "replicated_rule"
~~~

The following messages appear when logging in with some initiator. These 3 messages repeat several times.

~~~
<...>
2018-11-20 10:34:19.229 98 [DEBUG_SCSI_CMD] tcmu_print_cdb_info:1195 rbd/rbd.disk_2: 12 0 0 0 24 0 
2018-11-20 10:34:19.230 98 [WARN] tcmu_get_alua_grp:225 rbd/rbd.disk_2: Unsupported alua_access_type: Implicit and Explicit failover not supported.
2018-11-20 10:34:19.230 98 [ERROR] tcmu_get_alua_grps:319 rbd/rbd.disk_2: Could not get alua group ao.
<...>
~~~

Some other informations

~~~
$ cat /sys/kernel/config/target/core/user_1/rbd.disk_2/alua/default_tg_pt_gp/alua_access_type
Implicit and Explicit
$ cat /sys/kernel/config/target/core/user_1/rbd.disk_2/alua/ao/alua_access_type 
Implicit and Explicit
~~~

~~~
tcmu-runner.x86_64                    1.4.0-1.el7                  @tcmu-runner
ceph-iscsi-cli.noarch                 2.7-2.7.el7                  @ceph-iscsi  
ceph-iscsi-config.noarch              2.6-2.6.el7                  @ceph-iscsi  
~~~

**Comment by mikechristie** [comment]
2018-11-20 10:34:19.230 98 [WARN] tcmu_get_alua_grp:225 rbd/rbd.disk_2: Unsupported alua_access_type: Implicit and Explicit failover not supported.

Where did you get the ceph-iscsi-config rpm?

You need to update it. There was a issue where ceph-iscsi-config released a version expecting tcmu-runner to have a new feature but in the end runner 1.4 did not have it.

Did you get the rims from shamen? If so there is a new version

https://1.chacra.ceph.com/r/ceph-iscsi-config/master/179cbf423b89c1a875249f263c9f22f9e8314265/centos/7/flavors/default/noarch/

**Comment by Venator-Fox** [comment]
I updated the package in all containers via the linked rpm you provided. The luns now appear on all initators.  
THANK YOU for all your assistance on this. Can I send you some food.

I think I saw you mention this on some mailing lists when digging earlier, around version issues?  
[ceph-users tcmu iscsi](http://lists.ceph.com/pipermail/ceph-users-ceph.com/2018-October/030532.html)  
[Issue 369 on open-iscsi/tcmu-runner](https://github.com/open-iscsi/tcmu-runner/issues/369#issuecomment-439246839)  

Sorry I didn't mention that earlier. I was unsure on those because I thought the container would have the rpms? This is the container with the `ceph/daemon:latest-mimic` tag from Docker Hub automated build 8 days ago. I just used that instead of building. Everything is up to date when performing a `yum update` inside the tcmu_runner container. I however do not see any ceph-iscsi-config strings when searching the repos in the container.

Once again, thanks! Both on this and for #1220!

**Comment by mikechristie** [comment]
> THANK YOU for all your assistance on this. Can I send you some food.

No problem. Do not worry about it. Thanks for hanging in so long.

I will dig into the build and repo issue so we can hopefully get it fixed for future users. Thanks for the investigative work. It gave me a good start.

**Comment by vondrt4** [comment]
Thank you! This helped me as well, although I'm not building containers but a VM. The documentation really needs to be more verbose on where to get which packages.

I first started from Centos, but after I realized I'd have to pull in half of Fedora to get all the dependencies, I gave up.

Then I started again from Fedora, pulled in a quarter of Centos, and got stuck on weird kernel errors. The culprit was tcmu-runner, which is available in Fedora, but in an older version.The packages from shaman do not have minimum versions of dependencies set on them! 

All in all, the whole experience remindes me of the good days of Slackware on my 486. You had to hunt for packages all over the net on shady sites and solve dependencies manually.

I'm glad it works and I'm not touching the monster I created until someone packages this for Debian or Ubuntu. Amen.
